tools/defrag_all_hashes.py

```diff
@@ -0,0 +1,100 @@
+import redis.asyncio as aioredis
+import argparse
+import asyncio
+
+"""
+This script iterates over all hash keys and "recycles" them.
+Recycling is done by DUMPing the hash first and then re-creating it with EXPIRE.
+This will trigger re-allocation of internal hash data structures in order to reduce
+memory fragmentation.
+"""
+
+SCRIPT = """
+local recycled = 0
+for _, key in ipairs(KEYS) do
+    local ttl = redis.call('PTTL', key)
+    local dumpedData = redis.call('DUMP', key)
+
+    if dumpedData then
+        redis.call('RESTORE', key, 0, dumpedData, 'REPLACE')
+        if ttl > 0 then
+            redis.call('PEXPIRE', key, ttl)
+        end
+        recycled = recycled + 1
+    end
+end
+return recycled
+"""
+
+total_recycled = 0
+
+
+async def workerfn(client_supplier, sha, queue):
+    global total_recycled
+
+    r = client_supplier()
+    while True:
+        keys = await queue.get()
+
+        try:
+            recycled = await r.evalsha(sha, len(keys), *keys)
+        except Exception as e:
+            raise SystemExit(e)
+
+        if isinstance(recycled, int):
+            total_recycled += recycled
+        else:
+            print("Error recycling", recycled)
+
+        queue.task_done()
+
+
+async def infofn():
+    while True:
+        await asyncio.sleep(0.5)
+        print("Keys processed:", total_recycled)
+
+
+async def main(client_supplier, num_workers, queue_size, batch_size):
+    r = client_supplier()
+    sha = await r.script_load(SCRIPT)
+    queue = asyncio.Queue(maxsize=queue_size)
+
+    workers = [
+        asyncio.create_task(workerfn(client_supplier, sha, queue)) for _ in range(num_workers)
+    ]
+    info_worker = asyncio.create_task(infofn())
+
+    keys = []
+    async for key in r.scan_iter("*", count=batch_size * 2, type="hash"):
```

Comment:
there is nothing special for hashes here. I suggest making the type an optional argument and scan everything by default .
Also, let's rename the script to `defrag_db.py`

Reply:
Done